### PR TITLE
Have data accessing modules wait for evolutions to complete

### PR DESCRIPTION
**Pull request by gwendolyngoetz**
#### TL;DR
Make custom modules that edit the database wait for evolutions to complete and also finish running before the app starts servicing clients.

#### More
Custom modules making changes to the database that are expected to be done before the app is live relied on Akka's ActorSystem, but this would result in work being completed after the app is live. The change checks that the evolutions have completed before moving on.

I've been told this didn't work back in the early days of the project which is why Akka was used. Later the delay was added to avoid a race condition. This was [considered hacky at the time](https://github.com/civiform/civiform/pull/3663).

> [!NOTE]
>The [recommended approach from Play](https://www.playframework.com/documentation/3.0.x/Migration27#Application-starts-when-evolutions-scripts-need-to-be-applied-in-DEV-mode) is to inject `ApplicationEvolutions` and check that `upToDate` is true before doing anything. This is true both in the 2.7 migration guide and what you find Play contributors say in their repo. 
> > That’s why we added [ApplicationEvolutions.upToDate](https://www.playframework.com/documentation/3.0.x/api/scala/play/api/db/evolutions/ApplicationEvolutions.html#upToDate:Boolean) which indicates if the process of applying evolutions is finished or not. Only if that method returns true you can be sure that all evolutions scripts were executed successfully. upToDate will return true at some point eventually, because each time you apply or resolve evolutions scripts in DEV mode an application automatically restarts, re-initializing all it’s modules.

This is in the migration guide, but hasn't made it into the [docs](https://github.com/playframework/playframework/issues/9910) yet. There was a backport a while ago [changing the DEV mode part](https://github.com/playframework/playframework/pull/10030), but the checking upToDate remains true.

Additional notes can be found in the [Slack convo](https://civiform.slack.com/archives/C01QKN2UNMA/p1722902344358089).

#### Log output of start the app now (trimmed, logging was added for testing and is not in the final changes)
```
INFO  p.a.h.HttpErrorHandlerExceptions - Registering exception handler: guice-provision-exception-handler
INFO  m.EsriModule - Using services.geo.esri.RealEsriClient class for Esri client
WARN  m.SettingsMigrationModule - SettingsMigrationModule - ctor
INFO  m.SecurityModule - Using ADFS for admin auth provider
INFO  m.SecurityModule - Using generic OIDC for applicant auth provider
WARN  m.DatabaseSeedModule - DatabaseSeedModule - ctor
WARN  m.DurableJobModule - DurableJobModule - ctor
INFO  p.a.d.DefaultDBApi - Database [default] initialized
INFO  p.a.d.HikariCPConnectionPool - Creating Pool for datasource 'default'
INFO  io.ebean - ebean version: unknown
INFO  io.ebean.core - Started database[default] platform[POSTGRES] in 232ms

DEBUG p.a.d.e.DefaultEvolutionsApi - Execute
DEBUG p.a.d.e.DefaultEvolutionsApi - Finished in 4ms
      ... lots more evolutions ...
DEBUG p.a.d.e.DefaultEvolutionsApi - Execute
DEBUG p.a.d.e.DefaultEvolutionsApi - Finished in 0ms

WARN  m.SettingsMigrationModule - SettingsMigrationModule.SettingsMigrator - ctor
TRACE m.SettingsMigrationModule - SettingsMigrationModule.SettingsMigrator - scheduleOnce - start
DEBUG io.ebean.SQL - txn[]
INFO  s.s.SettingsService - Migrated 24 settings from config to database.
TRACE m.SettingsMigrationModule - SettingsMigrationModule.SettingsMigrator - scheduleOnce - end

WARN  m.DatabaseSeedModule - DatabaseSeedModule.DatabaseSeedScheduler - ctor
TRACE m.DatabaseSeedModule - DatabaseSeedModule.DatabaseSeedScheduler - scheduleOnce - start
INFO  s.s.DatabaseSeedTask - DatabaseSeekTask - run
DEBUG io.ebean.SQL - txn[]
INFO  s.s.DatabaseSeedTask - Created canonical question "Name"
DEBUG io.ebean.SQL - txn[]
INFO  s.s.DatabaseSeedTask - Created canonical question "Applicant Date of Birth"

WARN  m.DurableJobModule - DurableJobModule.DurableJobRunnerScheduler - ctor
TRACE m.DurableJobModule - DurableJobModule.DurableJobRunnerScheduler - beforeStartup - start
TRACE m.DurableJobModule - DurableJobModule.DurableJobRunnerScheduler - work.....
TRACE m.DurableJobModule - DurableJobModule.DurableJobRunnerScheduler - beforeStartup - end

INFO  play.api.Play - Application started (Dev) (no global state)
INFO  loggingfilter - GET       /       15ms    303 -> /callback?client_name=GuestClient

TRACE m.DurableJobModule - DurableJobModule.DurableJobRunnerScheduler - scheduleAtFixedRate - start
DEBUG io.ebean.SQL - txn[]
INFO  d.DurableJobRunner - JobRunner_Start thread ID=446
DEBUG io.ebean.SQL - txn[]
INFO  d.DurableJobRunner - JobRunner_Stop thread_ID=446
TRACE m.DurableJobModule - DurableJobModule.DurableJobRunnerScheduler - scheduleAtFixedRate - end

TRACE m.DurableJobModule - DurableJobModule.DurableJobRunnerScheduler - scheduleAtFixedRate - start
DEBUG io.ebean.SQL - txn[]
INFO  d.DurableJobRunner - JobRunner_Start thread ID=446
DEBUG io.ebean.SQL - txn[]
INFO  d.DurableJobRunner - JobRunner_Stop thread_ID=446
TRACE m.DurableJobModule - DurableJobModule.DurableJobRunnerScheduler - scheduleAtFixedRate - end
```

### Release Notes

Optimizing application startup

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label: < feature | enhancement | bug | under-development | dependencies | infrastructure | ignore-for-release | database >
- [x] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)

### Issue(s) this completes

Related to #3663

### CODE REVIEW

File: server/app/modules/DatabaseSeedModule.java [hunk 1]
```diff
@@ -23,17 +21,12 @@ public static final class DatabaseSeedScheduler {
...
-      actorSystem
-          .scheduler()
-          .scheduleOnce(
-              // schedule seed task for 5 sec from now. There is a race condition
-              // with Play evolutions. Evolutions must run before we seed database.
-              // It doesn't seem to be a way to run code after evolutions so just
-              // give them few sec to run.
-              Duration.ofSeconds(5), () -> databaseSeedTaskProvider.get().run(), executionContext);
+
+      if (applicationEvolutions.upToDate()) {
```
[the remark this answers is not in the document]
Comment (gwendolyngoetz) [reply server/app/modules/DatabaseSeedModule.java:27]: The note section above has more details and links. According to Matthias over with Play modules get reset each time an evolution is applied so we don't need to do any polling. Play will handle it.

[the remark this answers is not in the document]
Comment (rockycodes) [reply server/app/modules/DatabaseSeedModule.java:27]: I think this might only apply to dev. I was confused about the lack of mentions about production so I looked into the docs and found this bit.
```
When evolutions are activated, Play will check your database schema state before each request in DEV mode, or before starting the application in PROD mode. In DEV mode, if your database schema is not up to date, an error page will suggest that you synchronize your database schema by running the appropriate SQL script.
```
If I'm understanding this correctly, this race condition would not be an issue in production since prod mode waits for all the evolutions to be applied to start the app (as opposed to dev mode which starts the app... whenever? unless you have this `upToDate` check applied).

**Review by rockycodes — APPROVED** [review]
This looks fine to me and matches my understanding of the intended use from the docs. I tested it out locally and everything runs in order, the questions seed, etc. From my understanding of the docs, this was only an issue in dev so it shouldn't affect prod. But I would feel better if we could test it in prod mode (maybe just merging and manually deploying staging to make sure all is well).

File: server/app/modules/DatabaseSeedModule.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rockycodes) [reply server/app/modules/DatabaseSeedModule.java:27]: Would you add a comment or a bit in our wiki that explains why we need the `upToDate` method and/or links to that part of the migration guide?

**Review by dkatzz — APPROVED** [review]
Tested locally and LGTM as well, but agreed on testing in PROD mode and also about adding docs about this. Thanks for making this update!!

File: server/app/modules/DatabaseSeedModule.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (gwendolyngoetz) [reply server/app/modules/DatabaseSeedModule.java:39]: @rockycodes I added a small comment with a link to the PR.

**Comment by gwendolyngoetz** [comment]
Pushed to the Seattle staging site and had no issues. Logging shows the expected output, so I think we're good with this change.

I'm going to keep the logging lines in there, at least for now. They are trace level and can be enabled in the logback.xml if additional checking is needed.

**Review by rockycodes — APPROVED** [review]

**Review by dkatzz — APPROVED** [review]
